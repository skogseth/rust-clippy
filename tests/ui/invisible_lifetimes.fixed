#![allow(unused)]
#![warn(clippy::invisible_lifetimes)]

fn main() {}

// Struct with generic lifetime
struct MyStruct<'a>(&'a str);

fn input_value(_: MyStruct) {}

fn return_value() -> MyStruct<'static> {
    MyStruct("Hello")
}

fn identity(s: MyStruct) -> MyStruct {
    s
}
