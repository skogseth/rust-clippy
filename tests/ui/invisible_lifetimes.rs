#![allow(unused)]
#![warn(clippy::invisible_lifetimes)]

fn main() {}

// Struct with generic lifetime
struct SecretStr<'a>(&'a str);

fn input_value(_: SecretStr) {}

fn identity(s: SecretStr) -> SecretStr {
    s
}

// We don't want to activate on references
mod false_positive {
    fn input_value(_: &i32) {}

    fn identity(i: &i32) -> &i32 {
        i
    }
}

// Input and output are associated, but this is not clear!
// Want to lint on this and suggest <'_>
fn make_secret(s: &str) -> SecretStr {
    SecretStr(s)
}

// Input and output are associated, but the relationship is not clear!
// Want to lint on this and suggest <'_>
fn reveal_secret(s: SecretStr) -> &str {
    s.0
}

// Input and output is not associated!
// Want to lint on this and suggest <'static>?
fn confusing_lifetimes(s: &str) -> SecretStr {
    let _ = s; // Imagine we do something with this
    SecretStr("Hello world")
}
