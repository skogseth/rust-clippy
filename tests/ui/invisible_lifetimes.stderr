error[E0106]: missing lifetime specifier
  --> tests/ui/invisible_lifetimes.rs:11:22
   |
LL | fn return_value() -> MyStruct {
   |                      ^^^^^^^^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime, but this is uncommon unless you're returning a borrowed value from a `const` or a `static`, or if you will only have owned values
   |
LL | fn return_value() -> MyStruct<'static> {
   |                              +++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0106`.
